Use the console instead of a dedicated window when pressing chat/command keys

ShadowNinja opened the pull request:
Heavily rebased and enhanced version of @EXio4's old pull request.

This version makes the window only cover 20% of the screen if it's launched with `keymap_chat` or `keymap_command`.  Opening it with `keymap_console` opens it up to full size.

![screenshot_20160227_170942](https://cloud.githubusercontent.com/assets/3399154/13375723/330c7b10-dd75-11e5-92ac-3d2c54889365.png)

ShadowNinja commented:
@srifqi If you could in the text box you should be able to in the console.  I've never been able to use those functions anywhere though.

VanessaE commented:
Can't be merged until copy&paste and mouse capture (well, release) is sorted out on Linux.  With the old command line, when you open it, you can at least Ctrl-A, Ctrl-C and copy out from there to somewhere else (e.g. a text editor), and the mouse is released.  

With the console, neither of these work - you can't Ctrl-A to highlight, nor do shift-arrows work.

nerzhul commented:
Jenkins build passed

est31 commented:
Also, the f10 console is quite bad to use on android. I'll +1 if the pointer locking is disabled for the t and / versions of the console, and if it is either disabled on android, or there is an additional button to close the console.

ShadowNinja commented:
Copy/paste is fixed.  The pointer unlocking is just a matter of adding the code.  Ctrl-A should be simple to implement, although shift arrows may be a bit harder.

est31 commented:
Shift arrows do work in fact.

ShadowNinja commented:
@est31: In console?  No.
I've added code to handle selection and copying now though.
Also, I've made the console unlock the cursor, and I've made the console opened by F10 open to the full height of the window so that you have more space for history reading.
I think all issues with this PR are fixed now -- copy and paste works perfectly for me now.

everamzah commented:
What about turning on ESC_PAUSE_MODE, or such, so that the FPS is dropped (per the setting, default 20) when the console is dropped?  It would fix https://github.com/minetest/minetest/issues/3618 for me.

I hate to ask but in another game with a console, the height is also a setting.

Fixer-007 commented:
Please add time stamps (requested by me in https://github.com/minetest/minetest/issues/3716)

ShadowNinja commented:
@Fixer-007: Perhaps as a separate PR.

est31 commented on src/guiChatConsole.cpp:
```diff
@@ -362,13 +377,15 @@ void GUIChatConsole::drawPrompt()
-				y + m_fontsize.Y);
+				y + m_fontsize.Y * (1.0 - m_cursor_height),
+				x + m_fontsize.X * MYMAX(cursor_len, 1),
+				y + m_fontsize.Y * (cursor_len ? m_cursor_height+1 : 1)
```
space around +

est31 commented on src/guiChatConsole.cpp:
```diff
@@ -500,12 +514,32 @@ bool GUIChatConsole::OnEvent(const SEvent& event)
+			// Copy text to clipboard
+			if (prompt.getCursorLength() <= 0)
+				return true;
+			std::string selected = wide_to_narrow(prompt.getSelection());
```
can you use `wide_to_utf8` here? narrow narrows down things :)

ShadowNinja replied:
I'm pretty sure the system clipboard is supposed to be set according to the system's default encoding.

est31 replied:
I've tried to copy non ASCII stuff from the terminal and it seems to output broken text, while copying from a formspec text box gives the correct text.

I then replaced with `wide_to_utf8`, still not working. Strange...

est31 commented on src/guiChatConsole.cpp:
```diff
@@ -513,17 +547,14 @@ bool GUIChatConsole::OnEvent(const SEvent& event)
-				std::wstring wtext = narrow_to_wide(text);
-				m_chat_backend->getPrompt().input(wtext);
-			}
+				prompt.input(narrow_to_wide(text));
```
here too

est31 commented:
See the three things pointed out above otherwise looks good +1

est31 commented:
Okay tried the PR, some points:

* Clipboard issues, see above
* I think the "chat bar" flies in too slowly. I tested it on low FPS, is it FPS dependent? Can we live without a fly animation?
* I like how the mouse is free now
* If you go to the end of a string and then select with shift+left, the last character of the string wont be selected.
* Android must work obviously, does it?

ShadowNinja commented:
@est31
 * Actually the clipboard is apparently encoded in CP-1252.  I've fixed the chat console to use that encoding, but it only supports 255 chars.  Irrlicht will have to be fixed for proper wide char support (the text boxes are broken too).
 * It's a matter of taste, but I haven't touched the animation code.
 * Good :-)
 * Fixed.
 * I've added a separate chat window for Android.  It's basically the same as before except you don't need to press "Proceed" after already submitting your message.

I think this PR is ready to merge now.

est31 commented:
Okay from me as well :)

ShadowNinja commented:
effa24737d6997dc40508533635accca3ed099e9, 3edb7575a1ccca631994c0f5515b770b7a3e20d2, 430929e75a4e6c81d7dd2746570f925bb45e674e, 9dd38cf9682539ff5a2cf8e698a269cbcde3961a, 8b006a154b31a2fd0fb58e77db25debef4bb61e0

0-afflatus commented:
I am loving the behaviour of this, although copy-paste still doesn't work for me on Linux. I assume I've missed something, or need to enable some compile option. Does this depend on ncurses?

est31 commented:
@0-afflatus no, copying from minetest to an external application relies on irrlicht. You need most recent SVN version for it to work. There is no irrlicht release yet with the fix applied.

0-afflatus commented:
@est31 ok, understood.

The one issue that I noticed last night is that the console often opens several miles back in the conversation - it should open at the tail. Should I raise a separate issue for this?

est31 commented:
Perhaps its already tracked in the issue tracker I'm not sure, but yeah it would be not fault of irrlicht. Either way I don't really know whether remembering the scroll position if you aren't at the lowermost position its a feature or a bug.

0-afflatus commented:
It isn't remembering the scroll position, so that's a bug AFAIC. I will check through existing issues properly when not so brain-dead. (big in-game disciplinary hearing last night)
